langchain/src/experimental/openai_assistant/openai_assistant.ts

```diff
@@ -0,0 +1,278 @@
+import { type ClientOptions, OpenAI as OpenAIClient } from "openai";
+import { Run } from "openai/resources/beta/threads/index";
+import { AssistantCreateParams } from "openai/resources/beta/index.mjs";
+import { Runnable } from "../../schema/runnable/base.js";
+import { sleep } from "../../util/time.js";
+import { RunnableConfig } from "../../schema/runnable/config.js";
+import {
+  OutputType,
+  OpenAIAssistantFinish,
+  OpenAIAssistantAction,
+  OpenAIToolType,
+} from "./schema.js";
+import { StructuredTool } from "../../tools/base.js";
+import { formatToOpenAIFunction } from "../../tools/convert_to_openai.js";
+
+interface OpenAIAssistantRunnableInput {
+  client?: OpenAIClient;
+  clientOptions?: ClientOptions;
+  assistantId: string;
+  asAgent?: boolean;
+  pollIntervalMs?: number;
+}
+
+export class OpenAIAssistantRunnable<
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  RunInput extends Record<string, any>,
+  RunOutput extends OutputType
+> extends Runnable<RunInput, RunOutput> {
+  lc_namespace = ["langchain", "beta", "openai_assistant"];
+
+  private client: OpenAIClient;
+
+  assistantId: string;
+
+  pollIntervalMs = 1000;
+
+  asAgent = false;
+
+  constructor(fields: OpenAIAssistantRunnableInput) {
+    super();
+    this.client = fields.client || new OpenAIClient(fields?.clientOptions);
+    this.assistantId = fields.assistantId;
+    this.asAgent = fields.asAgent ?? false;
+  }
+
+  static async create<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunInput extends Record<string, any>,
+    RunOutput extends OutputType
+  >({
+    model,
+    name,
+    instructions,
+    tools,
+    client,
+    clientOptions,
+    asAgent,
+    pollIntervalMs,
+  }: Omit<OpenAIAssistantRunnableInput, "assistantId"> & {
+    model: string;
+    name?: string;
+    instructions?: string;
+    tools?: OpenAIToolType | Array<StructuredTool>;
+  }) {
+    const formattedTools =
+      tools?.map((tool) => {
+        // eslint-disable-next-line no-instanceof/no-instanceof
+        if (tool instanceof StructuredTool) {
+          return {
+            type: "function",
+            function: formatToOpenAIFunction(tool),
+          } as AssistantCreateParams.AssistantToolsFunction;
```

Comment:
Are the SDK types just incorrect or are we missing something?

Reply:
Judging by below I would make a separate method and verify the `AssistantToolsFunction` type